docs/the-new-architecture/pillars-turbomodule.md

```diff
@@ -3,16 +3,700 @@ id: pillars-turbomodules
 
-- JS spec (with all the supported features)
-- Configuration (package.json, cocoapods, gradle, …) and CodeGen
-- Native code (one section for iOS and one for Android)
-- Integration in an App (`yarn add` and how to connect the JS specs to the app itself)
-- Troubleshooting (common issues and how to solve them)
+This guide will show you how to create a basic TurboModule.
+
+:::caution
+TurboModules only work with the **New Architecture** enabled.
+To migrate to the **New Architecture**, follow the [Migration guide](../new-architecture-intro)
+:::
+
+## How to Create a TurboModule
+
+To create a TurboModule, we need to:
+
+1. Define a set of JavaScript specifications.
+2. Configure the module and inspect the code created by Codegen.
+3. Write the native code to finish implementing the module.
+
+## 1. Folder Setup
+
+In order to keep the module decoupled from the app, it's a good idea to define the module separately from the app, and then add it as a dependency to your app later.
+
+Next to your application, create a folder called `RTNCalculator`. (**RTN** stands for "**R**eact**T** **N**ative", and is a standard prefix for React Native modules).
+
+Within `RTNCalculator`, create three subfolders: `js`, `ios`, and `android`.
+
+The final result should look like this:
+
+```sh
+TurboModulesGuide
+├── MyApp
+└── RTNCalculator
+    ├── android
+    ├── ios
+    └── js
+```
+
+## 2. JavaScript Specification
+
+The **New Architecture** requires interfaces specified in a typed dialect of JavaScript (either [Flow](https://flow.org/) or [TypeScript](https://www.typescriptlang.org/)). **Codegen** will use these specifications to generate code in strongly-typed languages, including C++, Objective-C++, and Java.
+
+There are two requirements the file containing this specification must meet:
+
+1. The file **must** be named `Native<MODULE_NAME>`, with a `.js` or `.jsx` extension when using Flow, or a `.ts`, or `.tsx` extension when using TypeScript. Codegen will only look for files matching this pattern.
+2. The file must export a `TurboModuleRegistrySpec` object.
+
+<Tabs groupId="turbomodule-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
+<TabItem value="flow">
+
+```typescript title="NativeCalculator.js"
+// @flow
+import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
+import { TurboModuleRegistry } from 'react-native';
+
+export interface Spec extends TurboModule {
+  add(a: number, b: number): Promise<number>;
+}
+export default (TurboModuleRegistry.get<Spec>(
+  'RTNCalculator'
+): ?Spec);
+```
+
+</TabItem>
+<TabItem value="typescript">
+
+```typescript title="NativeCalculator.ts"
+import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
+import { TurboModuleRegistry } from 'react-native';
+
+export interface Spec extends TurboModule {
+  add(a: number, b: number): Promise<number>;
+}
+
+export default (TurboModuleRegistry.get<Spec>(
```

Comment:
Just a point of curiosity: Does exporting default like this defeat lazy loading? Is lazy loading obtained a different way? Or is that lazy loading feature entirely on the native side? (I assume not or I'd expect `TurboModuleRegistry.get` to return a promise)

Reply:
From prior experience with things like this, my understanding is that `get` would be called as soon as this was imported in another file, so in the example app it's not being lazily loaded. We'd either have to export a function that calls `get` or conditionally `require` this export to have lazy loading, correct?